---
title: "Figure4"
author: "Jacob T. Nearing and Gavin Douglas"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(reshape2)
library(pheatmap)
library(plyr)
library(cowplot)
library(ggplotify)
library(RColorBrewer)
library(ggbeeswarm)

### Folders where final figures are written
### Make sure this is pointed to your home directory
display_items_out <- "/home/gavin/github_repos/hackathon/Comparison_of_DA_microbiome_methods/Display_items/"

Data_set_names <- c(ArcticFireSoils="Soil - Fires",
                         ArcticFreshWater="Freshwater - Arctic",
                         ArcticTransects="Soil - Arctic",
                         art_scher="Human - RA",
                         asd_son= "Human - ASD",
                         BISCUIT= "Human - CD (1)",
                         Blueberry= "Soil - Blueberry",
                         Schubert="Human - C. diff (1)",
                         cdi_vincent="Human - C. diff (2)",
                         Chemerin="Mouse Facilities",
                         crc_baxter="Human - CC (1)",
                         crc_zeller="Human - CC (2)",
                         edd_singh="Human - Inf.",
                         Exercise="Mouse - Exercised",
                         glass_plastic_oberbeckmann="Marine - Plastic (4)",
                         GWMC_ASIA_NA="WWSR - Continents",
                         GWMC_HOT_COLD="WWSR - Temp.",
                         hiv_dinh="Human - HIV (1)",
                         hiv_lozupone="Human - HIV (2)",
                         Hiv_noguerajulian="Human - HIV (3)",
                         ibd_gevers="Human - CD (2)",
                         ibd_papa="Human - IBD",
                         Ji_WTP_DS="Freshwater - Treat.",
                         MALL="Human - ALL",
                         Goodrich="Human - OB (1)",
                         ob_ross="Human - OB (2)",
                         ob_turnbaugh="Human - OB (3)",
                         ob_zhu="Human - OB (4)",
                         ###"ob_zupancic",
                         Office="Built - Office",
                         par_scheperjans="Human - Par.",
                         sed_plastic_hoellein="Marine - Plastic (2)",
                         sed_plastic_rosato="Marine - Plastic (5)",
                         seston_plastic_mccormick="River - Plastic",
                         sw_plastic_frere="Marine - Plastic (1)",
                         Sw_sed_detender="Marine - Sediment",
                          t1d_alkanani="Human - T1D (1)",
                         t1d_mejialeon="Human - T1D (2)",
                         wood_plastic_kesy="Marine - Plastic (3)")

tool_names <- c(aldex2="ALDEx2", ancom="ANCOM-II", corncob="corncob", deseq2="DESeq2", edger="edgeR", lefse="LEfSe", 
                limma_voom_TMM="limma voom (TMM)", limma_voom_TMMwsp="limma voom (TMMwsp)", maaslin2="MaAsLin2",
                maaslin2rare="MaAsLin2 (rare)", metagenomeSeq="metagenomeSeq", ttestrare="t-test (rare)", 
                wilcoxonclr="Wilcoxon (CLR)", wilcoxonrare="Wilcoxon (rare)")


```

# Introduction

The below code parses simulation replicates where we do not expect to see differential abundance differences; the groups being compared are just randomly sampled controls. Heatmaps and boxplots are generated to summarize the data. Note that despite this script's name, the boxplots are now in the main text and the heatmap is in the supplement, which is a choice we made during revisions.

# Analyses

# Creating summary dataframes

```{r}

get_total_feats <- function(df_list, rare=FALSE){
  
  rare_tools <- c("lefse", "maaslin2rare", "ttestrare", "wilcoxonrare")
  
  if(!rare){
    total_feats <- nrow(df_list[[1]])
  }else{
    total_feats <- colSums(!is.na(df_list[[1]][, rare_tools]))
  }
  
  return(total_feats)
}

### get percent significant for each tool
get_percent_sig <- function(df_list){
  
  rare_tools <- c("lefse", "maaslin2rare", "ttestrare", "wilcoxonrare")
  total_feats <- nrow(df_list[[1]])
  
  #since each of these tools should return a value if they are tested we can get the number of tested features counting 
  # the number of non-NA values
  rare_nonNA_val <- colSums(!is.na(df_list[[1]][, rare_tools]))
  
  
  ### first convert DF to binary 1 and 0 representing sig and nonsig features.
  temp_sig_list <- list()
  for (i in 1:length(df_list)){
    temp_tab <- df_list[[i]]
    temp_tab[temp_tab < 0.05] <- 2
    temp_tab[temp_tab != 2] <- 0
    temp_tab[temp_tab == 2] <- 1
    temp_sig_list[[i]] <- temp_tab
  }
  
  ### now we need to go through these and convert to % significant found for each tool
  temp_per_sig <- list()
  temp_raw_count <- list()
  
  
  ## go through each data frame in the list and get the total number of significant hits along with
  ## the % of significant hits found
  for(i in 1:length(temp_sig_list)){
    
    #get # of significant features for each tool
    num_hits <- colSums(temp_sig_list[[i]], na.rm = T)
    ## save this to a list that we will later bind into a dataframe.
    temp_raw_count[[i]] <- num_hits 
    
    ### now this part is tricky b/c some tools use rarified data and some do not...
    ## this will be the # representing total features tested
    ## the problem is to get this data we would need to load in the large list so we will manually
    for(j in 1:length(num_hits)){
      #check if it is rarified tool
      if(names(num_hits)[j]=="lefse"){
        num_hits[j] <- num_hits[j]/rare_nonNA_val[1]
      }else if(names(num_hits)[j]=="maaslin2rare"){
        num_hits[j] <- num_hits[j]/rare_nonNA_val[2]
      }else if(names(num_hits)[j]=="ttestrare"){
        num_hits[j] <- num_hits[j]/rare_nonNA_val[3]
      }else if(names(num_hits)[j]=="wilcoxonrare"){
        num_hits[j] <- num_hits[j]/rare_nonNA_val[4]
      }else{
           num_hits[j] <- num_hits[j]/total_feats
         }
           
      temp_per_sig[[i]] <- num_hits
      
    }
  }
  
  temp_per_sig_df <- rbind.fill(lapply(temp_per_sig,function(y){as.data.frame(t(y),stringsAsFactors=FALSE)}))
  temp_raw_count_df <- rbind.fill(lapply(temp_raw_count,function(y){as.data.frame(t(y),stringsAsFactors=FALSE)}))
  ret_data <- list()
  ret_data[['raw_count']] <- temp_raw_count_df
  ret_data[['per_sig']] <- temp_per_sig_df * 100
  return(ret_data)
  
}
```

## Filtered
```{r}
Filtered_FD_data <- readRDS("/home/jacob/GitHub_Repos/Hackathon_testing/Data/FD_data_filt_21_09_17.RDS")


ArcticFreshWaters_per_sig <- get_percent_sig(Filtered_FD_data[["ArcticFreshWaters"]])

Office_per_sig <- get_percent_sig(Filtered_FD_data[["Office"]])

Blueberry_per_sig <- get_percent_sig(Filtered_FD_data[["Blueberry"]])

ob_goodrich_per_sig <- get_percent_sig(Filtered_FD_data[["ob_goodrich"]])

hiv_noguerajulian_per_sig <- get_percent_sig(Filtered_FD_data[["hiv_noguerajulian"]])

Ji_WTP_DS_per_sig <- get_percent_sig(Filtered_FD_data[["Ji_WTP_DS"]])

cdi_schubert_per_sig <- get_percent_sig(Filtered_FD_data[["cdi_schubert"]])

sw_sed_detender_per_sig <- get_percent_sig(Filtered_FD_data[["sw_sed_detender"]])

## generete dataframe to create DF
Mean_filt_percent_sig <- data.frame(ArcticFreshWater = colMeans(ArcticFreshWaters_per_sig[[2]]),
                                    Office = colMeans(Office_per_sig[[2]]),
                                    Blueberry = colMeans(Blueberry_per_sig[[2]]),
                                    Goodrich = colMeans(ob_goodrich_per_sig[[2]]),
                                    Hiv_noguerajulian = colMeans(hiv_noguerajulian_per_sig[[2]]),
                                    Ji_WTP_DS = colMeans(Ji_WTP_DS_per_sig[[2]]),
                                    Schubert = colMeans(cdi_schubert_per_sig[[2]]),
                                    Sw_sed_detender = colMeans(sw_sed_detender_per_sig[[2]]))


Mean_raw_sig <- data.frame(ArcticFreshWater = colMeans(ArcticFreshWaters_per_sig[[1]]),
                                    Office = colMeans(Office_per_sig[[1]]),
                                    Blueberry = colMeans(Blueberry_per_sig[[1]]),
                                    Goodrich = colMeans(ob_goodrich_per_sig[[1]]),
                                    Hiv_noguerajulian = colMeans(hiv_noguerajulian_per_sig[[1]]),
                                    Ji_WTP_DS = colMeans(Ji_WTP_DS_per_sig[[1]]),
                                    Schubert = colMeans(cdi_schubert_per_sig[[1]]),
                                    Sw_sed_detender = colMeans(sw_sed_detender_per_sig[[1]]))

SD_filt_percent_sig <- data.frame(ArcticFreshWater = apply(ArcticFreshWaters_per_sig[[2]], 2, sd),
                                    Office = apply(Office_per_sig[[2]], 2, sd),
                                    Blueberry = apply(Blueberry_per_sig[[2]], 2, sd),
                                    Goodrich = apply(ob_goodrich_per_sig[[2]], 2, sd),
                                    Hiv_noguerajulian = apply(hiv_noguerajulian_per_sig[[2]], 2, sd),
                                    Ji_WTP_DS = apply(Ji_WTP_DS_per_sig[[2]], 2, sd),
                                    Schubert = apply(cdi_schubert_per_sig[[2]], 2, sd),
                                    Sw_sed_detender = apply(sw_sed_detender_per_sig[[2]], 2, sd))

Median_filt_percent_sig <- data.frame(ArcticFreshWater = apply(ArcticFreshWaters_per_sig[[2]], 2, median),
                                    Office = apply(Office_per_sig[[2]], 2, median),
                                    Blueberry = apply(Blueberry_per_sig[[2]], 2, median),
                                    Goodrich = apply(ob_goodrich_per_sig[[2]], 2, median),
                                    Hiv_noguerajulian = apply(hiv_noguerajulian_per_sig[[2]], 2, median),
                                    Ji_WTP_DS = apply(Ji_WTP_DS_per_sig[[2]], 2, median),
                                    Schubert = apply(cdi_schubert_per_sig[[2]], 2, median),
                                    Sw_sed_detender = apply(sw_sed_detender_per_sig[[2]], 2, median))


# Sort rownames
Mean_filt_percent_sig <- Mean_filt_percent_sig[sort(rownames(Mean_filt_percent_sig)), ]
Mean_raw_sig <- Mean_raw_sig[sort(rownames(Mean_raw_sig)), ]

rownames(Mean_filt_percent_sig) <- tool_names[rownames(Mean_filt_percent_sig)]
colnames(Mean_filt_percent_sig) <- Data_set_names[colnames(Mean_filt_percent_sig)]

breaksList = seq(0, 30, by = 0.2)

# Round values greater than 1
Mean_raw_sig[Mean_raw_sig >= 1] <- round(Mean_raw_sig[Mean_raw_sig >= 1])
Mean_raw_sig[Mean_raw_sig < 1] <- round(Mean_raw_sig[Mean_raw_sig < 1], 1)

# Order column names
Mean_raw_sig <- Mean_raw_sig[, order(colnames(Mean_filt_percent_sig))]
Mean_filt_percent_sig <- Mean_filt_percent_sig[, order(colnames(Mean_filt_percent_sig))]

filtered_heatmap_panel <- pheatmap::pheatmap(as.matrix(Mean_filt_percent_sig),
                                             cluster_cols = FALSE,
                                             cluster_rows = FALSE,
                                             display_numbers = round(Mean_raw_sig, 3),
                                             number_color = "Black",
                                             main = "Filtered",
                                             angle_col = 315,
                                             color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)),
                                             breaks = breaksList)


filtered_heatmap_panel

```


## Unfiltered
```{r}
Unfiltered_FD_data <- readRDS("/home/jacob/GitHub_Repos/Hackathon_testing/Data/FD_data_unfilt_21_04_08.RDS")

Unfiltered_FD_data_ext <- readRDS("/home/jacob/GitHub_Repos/Hackathon_testing/Data/FD_data_ext_21_04_08.RDS")

## Has 100 reps for all tools
ArcticFreshWater_FD_data_comb  <- c(Unfiltered_FD_data[["ArcticFreshWaters"]], Unfiltered_FD_data_ext[["ArcticFreshWaters"]])

#Has 100 reps for all tools
Blueberry_FD_data_comb <- c(Unfiltered_FD_data[["Blueberry"]], Unfiltered_FD_data_ext[["Blueberry"]])

#Has 100 reps for all tools
ob_goodrich_data_comb <- c(Unfiltered_FD_data[["ob_goodrich"]], Unfiltered_FD_data_ext[["ob_goodrich"]])

#Has 100 reps for all tools except ancom and corncob
hiv_noguerajulian_data_comb <- c(Unfiltered_FD_data[["hiv_noguerajulian"]], Unfiltered_FD_data_ext[["hiv_noguerajulian"]])

#All other datasets have 100 reps for all tools except ancom, aldex2 and corncob
Ji_WTP_DS_data_comb <- c(Unfiltered_FD_data[["Ji_WTP_DS"]], Unfiltered_FD_data_ext[["Ji_WTP_DS"]])

Office_data_comb <- c(Unfiltered_FD_data[["Office"]], Unfiltered_FD_data_ext[["Office"]])

cdi_schubert_data_comb <- c(Unfiltered_FD_data[["cdi_schubert"]], Unfiltered_FD_data_ext[["cdi_schubert"]])

sw_sed_detender_data_comb <- c(Unfiltered_FD_data[["sw_sed_detender"]], Unfiltered_FD_data_ext[["sw_sed_detender"]])



ArcticFreshWaters_per_sig_un <- get_percent_sig(ArcticFreshWater_FD_data_comb)


Blueberry_per_sig_un <- get_percent_sig(Blueberry_FD_data_comb)


ob_goodrich_per_sig_un <- get_percent_sig(ob_goodrich_data_comb)

hiv_noguerajulian_per_sig_un <- get_percent_sig(hiv_noguerajulian_data_comb)


Ji_WTP_DS_per_sig_un <- get_percent_sig(Ji_WTP_DS_data_comb)

Office_per_sig_un <- get_percent_sig(Office_data_comb)

cdi_schubert_per_sig_un <- get_percent_sig(cdi_schubert_data_comb)

sw_sed_detender_per_sig_un <- get_percent_sig(sw_sed_detender_data_comb)


Mean_nonfilt_percent_sig <- data.frame(ArcticFreshWater = colMeans(ArcticFreshWaters_per_sig_un[[2]], na.rm = T),
                                       Office = colMeans(Office_per_sig_un[[2]], na.rm=T),
                                       Blueberry = colMeans(Blueberry_per_sig_un[[2]], na.rm=T),
                                       Goodrich = colMeans(ob_goodrich_per_sig_un[[2]], na.rm=T),
                                       Hiv_noguerajulian = colMeans(hiv_noguerajulian_per_sig_un[[2]], na.rm=T),
                                       Ji_WTP_DS = colMeans(Ji_WTP_DS_per_sig_un[[2]], na.rm=T),
                                       Schubert = colMeans(cdi_schubert_per_sig_un[[2]], na.rm=T),
                                       Sw_sed_detender = colMeans(sw_sed_detender_per_sig_un[[2]], na.rm=T))


Mean_raw_nonfilt <- data.frame(ArcticFreshWater = colMeans(ArcticFreshWaters_per_sig_un[[1]], na.rm=T),
                               Office = colMeans(Office_per_sig_un[[1]], na.rm=T),
                                       Blueberry = colMeans(Blueberry_per_sig_un[[1]], na.rm=T),
                                       Goodrich = colMeans(ob_goodrich_per_sig_un[[1]], na.rm = T),
                                       Hiv_noguerajulian = colMeans(hiv_noguerajulian_per_sig_un[[1]], na.rm = T),
                                       Ji_WTP_DS = colMeans(Ji_WTP_DS_per_sig_un[[1]], na.rm=T),
                                       Schubert = colMeans(cdi_schubert_per_sig_un[[1]], na.rm=T),
                                       Sw_sed_detender = colMeans(sw_sed_detender_per_sig_un[[1]], na.rm=T))


# Sort rownames
Mean_nonfilt_percent_sig <- Mean_nonfilt_percent_sig[sort(rownames(Mean_nonfilt_percent_sig)), ]
Mean_raw_nonfilt <- Mean_raw_nonfilt[sort(rownames(Mean_raw_nonfilt)), ]

rownames(Mean_nonfilt_percent_sig) <- tool_names[rownames(Mean_nonfilt_percent_sig)]
colnames(Mean_nonfilt_percent_sig) <- Data_set_names[colnames(Mean_nonfilt_percent_sig)]


breaksList = seq(0, 30, by = 0.2)

# Round values greater than 1
Mean_raw_nonfilt[Mean_raw_nonfilt >= 1] <- round(Mean_raw_nonfilt[Mean_raw_nonfilt >= 1])
Mean_raw_nonfilt[Mean_raw_nonfilt < 1] <- round(Mean_raw_nonfilt[Mean_raw_nonfilt < 1], 1)

# Order column names
Mean_raw_nonfilt <- Mean_raw_nonfilt[, order(colnames(Mean_nonfilt_percent_sig))]
Mean_nonfilt_percent_sig <- Mean_nonfilt_percent_sig[, order(colnames(Mean_nonfilt_percent_sig))]


unfiltered_heatmap_panel <- pheatmap::pheatmap(Mean_nonfilt_percent_sig,
                                               cluster_cols=F,
                                               cluster_rows = F,
                                               display_numbers = round(Mean_raw_nonfilt, 3),
                                               number_color="Black",
                                               main="Unfiltered",
                                               angle_col=315,
                                               color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList)),
                                               breaks = breaksList)

unfiltered_heatmap_panel
```


# Final heatmap
Heatmap to be saved and then edited outside of R so that the colour key is labelled and only shown once for both plots.

```{r}
Figure4_plot_old <- cowplot::plot_grid(ggplotify::as.ggplot(unfiltered_heatmap_panel),
                                   ggplotify::as.ggplot(filtered_heatmap_panel),
                                   nrow=1, labels = c("A", "B"))

Figure4_plot_old

ggsave(filename=paste(display_items_out, "Misc_figures", "OLD_Figure4_FDR_heatmaps_raw.pdf", sep="/"),
       plot = Figure4_plot_old, width = 9, height=3.5, units="in", dpi=600)
```


# Main-text boxplot figures
This plots were originally supplementay but we have decided to include them as main figures.

```{r}
qual_col <- c("grey", #ALDEx2
              "palegreen3", # ANCOM
              "dodgerblue4", #corncob
              "darkorange", # DESeq2
              "seagreen4", # edgeR
              "ivory3", # LEfSe
              "darkorchid1", "darkorchid3", # limma-voom x 2
              "yellow1", "yellow3", # Maaslin 2
              "steelblue2", # metagenomeseq
              "magenta1", # t-test
              "coral2", "coral4") # Wilcoxon x 2

filt_percent_sig <- do.call(rbind, list(ArcticFreshWater = melt(ArcticFreshWaters_per_sig[[2]]),
                                         Office = melt(Office_per_sig[[2]]),
                                         Blueberry = melt(Blueberry_per_sig[[2]]),
                                         Goodrich = melt(ob_goodrich_per_sig[[2]]),
                                         Hiv_noguerajulian = melt(hiv_noguerajulian_per_sig[[2]]),
                                         Ji_WTP_DS = melt(Ji_WTP_DS_per_sig[[2]]),
                                         Schubert = melt(cdi_schubert_per_sig[[2]]),
                                         Sw_sed_detender = melt(sw_sed_detender_per_sig[[2]])))

filt_percent_sig$dataset <- gsub("\\..*$", "", rownames(filt_percent_sig))

colnames(filt_percent_sig) <- c("Tool", "Percent_sig", "Dataset")

filt_percent_sig$Tool <- as.character(filt_percent_sig$Tool)
filt_percent_sig$Dataset <- as.character(filt_percent_sig$Dataset)

filt_percent_sig$Tool <- factor(tool_names[filt_percent_sig$Tool])
filt_percent_sig$Tool <- factor(filt_percent_sig$Tool, levels = rev(levels(filt_percent_sig$Tool)))

#get total number of sig features
total_filt_feats <- lapply(Filtered_FD_data, get_total_feats)

names(total_filt_feats) <- c("Office", "ArcticFreshWater", "Blueberry", "Goodrich", "Hiv_noguerajulian",
                             "Ji_WTP_DS", "Schubert", "Sw_sed_detender")

matchs <- match(filt_percent_sig$Dataset, names(total_filt_feats))


filt_percent_sig$Dataset <- Data_set_names[filt_percent_sig$Dataset]

filt_percent_sig$Dataset_num <- paste(filt_percent_sig$Dataset, total_filt_feats[matchs], sep="\n  ASVs: ")


filt_percent_sig_boxplots <- ggplot(filt_percent_sig, aes(y = Tool, x = Percent_sig, fill = Tool)) +
                                    geom_boxplot(outlier.shape = NA, colour = "black") +
                                    geom_quasirandom(groupOnX = FALSE, size = 1, pch = 21, colour = "black") +
                                    theme_bw() +
                                    xlab(expression("Percent significant after multiple-test correction (log"[10]*"-scale)")) +
                                    ylab("") +
                                    theme(legend.position = "none") +
                                    theme(plot.title = element_text(hjust = 0.5)) +
                                    facet_wrap(. ~ Dataset_num, nrow = 2, labeller = labeller(type = label_parsed)) +
                                    scale_x_continuous(trans = scales::pseudo_log_trans(base = 10),
                                                       breaks = c(0, 1, 2, 5, 10, 30, 100),
                                                       limits = c(0, 100)) +
                                    scale_colour_manual(values = qual_col) +
                                    scale_fill_manual(values = qual_col) +
                                    ggtitle("Filtered")



unfilt_percent_sig <- do.call(rbind, list(ArcticFreshWater = melt(ArcticFreshWaters_per_sig_un[[2]]),
                                         Office = melt(Office_per_sig_un[[2]]),
                                         Blueberry = melt(Blueberry_per_sig_un[[2]]),
                                         Goodrich = melt(ob_goodrich_per_sig_un[[2]]),
                                         Hiv_noguerajulian = melt(hiv_noguerajulian_per_sig_un[[2]]),
                                         Ji_WTP_DS = melt(Ji_WTP_DS_per_sig_un[[2]]),
                                         Schubert = melt(cdi_schubert_per_sig_un[[2]]),
                                         Sw_sed_detender = melt(sw_sed_detender_per_sig_un[[2]])))

unfilt_percent_sig$dataset <- gsub("\\..*$", "", rownames(unfilt_percent_sig))

colnames(unfilt_percent_sig) <- c("Tool", "Percent_sig", "Dataset")

unfilt_percent_sig$Tool <- as.character(unfilt_percent_sig$Tool)
unfilt_percent_sig$Dataset <- as.character(unfilt_percent_sig$Dataset)

unfilt_percent_sig$Tool <- factor(tool_names[unfilt_percent_sig$Tool])
unfilt_percent_sig$Tool <- factor(unfilt_percent_sig$Tool, levels = rev(levels(unfilt_percent_sig$Tool)))

#get total number of ASVs tested
total_unfilt_feats <- lapply(Unfiltered_FD_data, get_total_feats)
names(total_unfilt_feats) <- c("ArcticFreshWater", "Blueberry", "Goodrich", "Hiv_noguerajulian", "Ji_WTP_DS", "Office", "Schubert", "Sw_sed_detender")

matchs_unfilt <- match(unfilt_percent_sig$Dataset, names(total_unfilt_feats))



unfilt_percent_sig$Dataset <- Data_set_names[unfilt_percent_sig$Dataset]

unfilt_percent_sig$Dataset_num <- paste(unfilt_percent_sig$Dataset, total_unfilt_feats[matchs_unfilt], sep="\n  ASVs: ")

unfilt_percent_sig <- unfilt_percent_sig[-which(is.na(unfilt_percent_sig$Percent_sig)), ]

unfilt_percent_sig_boxplots <- ggplot(unfilt_percent_sig, aes(y = Tool, x = Percent_sig, fill = Tool)) +
                                    geom_boxplot(outlier.shape = NA) +
                                    geom_quasirandom(color = "black", groupOnX = FALSE, size = 1, pch = 21,) +
                                    theme_bw() +
                                    xlab(expression("Percent significant after multiple-test correction (log"[10]*"-scale)")) +
                                    ylab("") +
                                    theme(legend.position = "none") +
                                    theme(plot.title = element_text(hjust = 0.5)) +
                                    facet_wrap(. ~ Dataset_num, nrow = 2) +
                                    scale_x_continuous(trans = scales::pseudo_log_trans(base = 10),
                                                       breaks = c(0, 1, 2, 5, 10, 30, 100),
                                                       limits = c(0, 100)) +
                                    scale_fill_manual(values = qual_col) +
                                    ggtitle("Unfiltered")

unfilt_percent_sig_boxplots_w_legend <- ggplot(unfilt_percent_sig, aes(y = Tool, x = Percent_sig, fill = Tool)) +
                                    geom_boxplot(outlier.shape = NA) +
                                    geom_quasirandom(color = "black", groupOnX = FALSE, size = 1, pch = 21,) +
                                    theme_bw() +
                                    xlab(expression("Percent significant after multiple-test correction (log"[10]*"-scale)")) +
                                    ylab("") +
                                    theme(plot.title = element_text(hjust = 0.5)) +
                                    facet_wrap(. ~ Dataset_num, nrow = 2) +
                                    scale_x_continuous(trans = scales::pseudo_log_trans(base = 10),
                                                       breaks = c(0, 1, 2, 5, 10, 30, 100),
                                                       limits = c(0, 100)) +
                                    scale_fill_manual(values = qual_col, guide = guide_legend(reverse = TRUE)) +
                                    ggtitle("Unfiltered")

boxplots_legend <- plot_grid(NULL, cowplot::get_legend(unfilt_percent_sig_boxplots_w_legend), NULL, nrow = 3, ncol = 1, rel_widths = c(1/3, 1/3, 1/3))



filt_percent_sig_boxplots

unfilt_percent_sig_boxplots

#write plotting data
write.table(x=unfilt_percent_sig_boxplots$data, file="~/GitHub_Repos/Clean_Hackathon/Plotting_data/Main_Figures/Figure4A_boxplots.csv", col.names = NA,
            row.names=T, quote=F, sep=",")

write.table(x=unfilt_percent_sig_boxplots$data, file="~/GitHub_Repos/Clean_Hackathon/Plotting_data/Main_Figures/Figure4B_boxplots.csv", col.names = NA,
            row.names=T, quote=F, sep=",")

New_Figure4_boxplots_no_legend <- plot_grid(unfilt_percent_sig_boxplots, filt_percent_sig_boxplots, labels=c("A", "B"), nrow=2)

New_Figure4 <- plot_grid(New_Figure4_boxplots_no_legend, boxplots_legend, ncol=2, nrow=1, rel_widths = c(0.8, 0.2))
New_Figure4


ggsave(filename=paste(display_items_out, "Main_figures", "Figure4.png", sep="/"),
       plot = New_Figure4, width = 10, height=10, units="in", dpi=300)

ggsave(filename=paste(display_items_out, "Main_figures", "Figure4.pdf", sep="/"),
       plot = New_Figure4, width = 10, height=10, units="in", dpi=600)


#old when boxplots were supplemental data... 
#ggsave(filename=paste(display_items_out, "Supp_figures", "Supp_FDR_filt_boxplots.png", sep="/"),
       #plot = filt_percent_sig_boxplots, width = 10, height=7, units="in", dpi=300)

#ggsave(filename=paste(display_items_out, "Supp_figures", "Supp_FDR_filt_boxplots.pdf", sep="/"),
       #plot = filt_percent_sig_boxplots, width = 10, height=7, units="in", dpi=600)

#ggsave(filename=paste(display_items_out, "Supp_figures", "Supp_FDR_unfilt_boxplots.png", sep="/"),
       #plot = unfilt_percent_sig_boxplots, width = 10, height=7, units="in", dpi=300)

#ggsave(filename=paste(display_items_out, "Supp_figures", "Supp_FDR_unfilt_boxplots.pdf", sep="/"),
       #plot = unfilt_percent_sig_boxplots, width = 10, height=7, units="in", dpi=600)

```

### Text summary stats

```{r}
print("edgeR filtered")
print("mean")
print(mean(as.numeric(Mean_filt_percent_sig["edgeR", ])))
print("sd")
print(sd(as.numeric(Mean_filt_percent_sig["edgeR", ])))

print("========")

print("LEfSe filtered")
print("mean")
print(mean(as.numeric(Mean_filt_percent_sig["LEfSe", ])))
print("sd")
print(sd(as.numeric(Mean_filt_percent_sig["LEfSe", ])))


print("=======")
print("filtered - all tool means")
print(rowMeans(Mean_filt_percent_sig))


print("======")
print("unfiltered limma voom results")
print(Mean_nonfilt_percent_sig[grep("limma voom", rownames(Mean_nonfilt_percent_sig)), ])
  
```
